src/analysis/normed_space/weak_dual.lean

```diff
@@ -182,3 +200,256 @@ begin
+    { to_fun := f,
+      map_add' := begin
+        intros z₁ z₂,
+        have key := f_lin.map_add z₁ z₂,
+        rw h_f_lin at key,
+        repeat {rwa linear_of_mem_closure_range_apply
+          f f_in_closure₀ _ at key, },
+      end,
+      map_smul' := begin
+        intros c z,
+        have key := f_lin.map_smul c z,
+        rw h_f_lin at key,
+        repeat {rwa linear_of_mem_closure_range_apply
+          f f_in_closure₀ _ at key, },
+      end,
+      cont := f_cont, } with hφ,
+  use φ,
+  refine ⟨_, rfl⟩,
+  dunfold weak_dual.polar,
+  dunfold polar,
+  intros z hz,
+  apply norm_eval_le_of_mem_closure_norm_eval_le z 1 f,
+  have ss : weak_dual.polar 𝕜 s ⊆ {x' : weak_dual 𝕜 E | ∥x' z∥ ≤ 1},
+  { intros x' hx',
+    simp only [weak_dual.polar, polar, mem_set_of_eq, preimage_set_of_eq] at hx',
+    exact hx' z hz, },
+  apply closure_mono (image_subset _ ss),
+  exact mem_closure_iff_cluster_pt.mpr hf,
+end
+
+/-- The image under `weak_dual.to_Pi : weak_dual 𝕜 E → Π (_ : E), 𝕜` of the polar `polar s` of
+a neighborhood `s` of the origin is compact if the field `𝕜` is a proper topological space. -/
+lemma image_polar_compact [proper_space 𝕜] {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) :
+  is_compact ((weak_dual.to_Pi 𝕜 E) '' (weak_dual.polar 𝕜 s)) :=
+begin
+  cases polar_bounded_of_nhds_zero 𝕜 s_nhd with c hc,
+  have ss : (weak_dual.to_Pi 𝕜 E) '' (weak_dual.polar 𝕜 s) ⊆
+    (set.pi (univ : set E) (λ z, (closed_ball (0 : 𝕜) (c * ∥ z ∥)))),
+  { intros f hf,
+    simp only [mem_image, exists_exists_and_eq_and] at hf,
+    rcases hf with ⟨x', hx', f_eq⟩,
+    simp only [mem_closed_ball, dist_zero_right, mem_univ_pi],
+    intros z,
+    have bd : ∥ x' z ∥ ≤ c * ∥ z ∥,
+    { apply (continuous_linear_map.le_op_norm x' z).trans
+        (mul_le_mul (hc x' hx') (le_refl ∥z∥) (norm_nonneg z) _),
+      have c_nn := hc 0 (zero_mem_polar 𝕜 s),
+      rwa norm_zero at c_nn, },
+    have eq : x' z = f z := congr_fun f_eq z,
+    rwa eq at bd, },
+  apply compact_of_is_closed_subset _ _ ss,
+  { apply is_compact_univ_pi,
+    exact λ z, proper_space.is_compact_closed_ball 0 _, },
+  exact image_polar_closed s_nhd,
+end
+
+end weak_dual.to_Pi_embedding
+
+/-- The Banach-Alaoglu theorem: the polar set of a neighborhood `s` of the origin in a
+normed space `E` is a compact subset of `weak_dual 𝕜 E`. -/
+theorem weak_dual.is_compact_polar [proper_space 𝕜] {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) :
+  is_compact (weak_dual.polar 𝕜 s) :=
+begin
+  apply (weak_dual.to_Pi_embedding 𝕜 E).is_compact_iff_is_compact_image.mpr,
+  exact weak_dual.to_Pi_embedding.image_polar_compact s_nhd,
+end
+
+/-- The Banach-Alaoglu theorem: closed balls of the dual of a normed space `E` over `ℝ` or `ℂ`
+are compact in the weak-star topology. -/
+theorem weak_dual.is_compact_closed_ball
+  {𝕜 : Type*} [is_R_or_C 𝕜] {E : Type*} [normed_group E] [normed_space 𝕜 E] (r : ℝ) (hr : 0 < r) :
+  is_compact {x' : weak_dual 𝕜 E | (∥ x'.to_normed_dual ∥ ≤ r)} :=
```

Comment:
Perhaps, to make the geometric point clear,
```suggestion
  is_compact (metric.closed_ball (0 : normed_space.dual 𝕜 E) r : set (weak_dual 𝕜 E)) :=
```
This statement would be less unwieldy if you open the namespaces `metric` and `normed_space`.

Reply:
This fails, with "failed to synthesize type class instance for ... `pseudo_metric_space (weak_dual 𝕜 E)`", meaning that it is trying to create the metric ball in the wrong type synonym. Sure there are some ways to tell Lean to use the ball of `dual 𝕜 E` as a subset of `weak_dual 𝕜 E`, but my solutions for this have not always been pretty. I think something like `weak_dual.to_normed_dual ⁻¹' (metric.closed_ball 0 r)` is not terrible, but it still slightly obscures the geometric point...

Reply:
Oh, but this allows to get rid of a clumsy rewrite-lemma I needed before!

Thank you!

Reply:
Oops, I should have tested it first!  One other way is
```lean
  is_compact (id (closed_ball 0 r : set (normed_space.dual 𝕜 E)) : set (weak_dual 𝕜 E))
```
but your current way also seems fine.